File: conformance/code/individual-go/internal/service/encryption_service.go
package service

import (
	"context"

	"individual/internal/clients"
	"individual/internal/common"
	"individual/internal/config"
	"individual/internal/models"

	"github.com/rs/zerolog/log"
)

// EncryptionService handles PII encryption/decryption
type EncryptionService interface {
	EncryptIndividual(ctx context.Context, individual *models.Individual) error
	DecryptIndividual(ctx context.Context, individual *models.Individual) error
	EncryptIndividuals(ctx context.Context, individuals []models.Individual) error
	DecryptIndividuals(ctx context.Context, individuals []models.Individual) error
	HashMobileNumber(mobileNumber string) (string, error)
}

type encryptionService struct {
	vaultClient clients.VaultClient
	config      *config.VaultConfig
	hmacSecret  []byte
}

// NewEncryptionService creates a new encryption service. hmacSecret is the pepper for the
// mobile-number blind index (HMAC-SHA256); it must match the secret the validator uses so a
// stored hash and a lookup hash agree.
func NewEncryptionService(vaultClient clients.VaultClient, config *config.VaultConfig, hmacSecret []byte) EncryptionService {
	return &encryptionService{
		vaultClient: vaultClient,
		config:      config,
		hmacSecret:  hmacSecret,
	}
}

// EncryptIndividual encrypts PII fields in individual
func (s *encryptionService) EncryptIndividual(ctx context.Context, individual *models.Individual) error {
	if !s.config.Enabled {
		log.Debug().Str("individualID", individual.ID).Msg("vault disabled — skipping encryption (mobile hash still computed)")
		// Still hash mobile number for search even without encryption
		if individual.MobileNumber != "" {
			hash, err := s.HashMobileNumber(individual.MobileNumber)
			if err != nil {
				return err
			}
			individual.HashedMobileNumber = hash
		}
		return nil
	}

	// Encrypt mobile number
	if individual.MobileNumber != "" {
		plaintext := individual.MobileNumber
		hash, err := s.HashMobileNumber(plaintext)
		if err != nil {
			return err
		}

		encrypted, err := s.vaultClient.Encrypt(ctx, plaintext, individual.TenantID)
		if err != nil {
			log.Error().Err(err).Ctx(ctx).Str("individualID", individual.ID).Str("field", "mobileNumber").Msg("vault encrypt failed")
			return common.ErrEncryption.WithContext(map[string]interface{}{
				"field": "mobileNumber",
				"error": err.Error(),
			})
		}

		individual.MobileNumber = encrypted
		individual.HashedMobileNumber = hash
	}

	// Encrypt alt contact number
	if individual.AltContactNumber != "" {
		encrypted, err := s.vaultClient.Encrypt(ctx, individual.AltContactNumber, individual.TenantID)
		if err != nil {
			log.Error().Err(err).Ctx(ctx).Str("individualID", individual.ID).Str("field", "altContactNumber").Msg("vault encrypt failed")
			return common.ErrEncryption
		}
		individual.AltContactNumber = encrypted
	}

	// Encrypt identifiers (especially Aadhaar)
	for i := range individual.Identifiers {
		if individual.Identifiers[i].IdentifierType == common.IdentifierTypeAadhaar && individual.Identifiers[i].IdentifierID != "" {
			encrypted, err := s.vaultClient.Encrypt(ctx, individual.Identifiers[i].IdentifierID, individual.TenantID)
			if err != nil {
				log.Error().Err(err).Ctx(ctx).
					Str("individualID", individual.ID).
					Str("identifierType", individual.Identifiers[i].IdentifierType).
					Msg("vault encrypt failed on identifier")
				return common.ErrEncryption.WithContext(map[string]interface{}{
					"field": "identifierId",
					"type":  individual.Identifiers[i].IdentifierType,
				})
			}
			individual.Identifiers[i].IdentifierID = encrypted
		}
	}

	return nil
}

// DecryptIndividual decrypts PII fields in individual
func (s *encryptionService) DecryptIndividual(ctx context.Context, individual *models.Individual) error {
	if !s.config.Enabled {
		log.Debug().Str("individualID", individual.ID).Msg("vault disabled — skipping decryption")
		return nil
	}

	if individual.MobileNumber != "" {
		decrypted, err := s.vaultClient.Decrypt(ctx, individual.MobileNumber, individual.TenantID)
		if err != nil {
			log.Error().Err(err).Ctx(ctx).Str("individualID", individual.ID).Str("field", "mobileNumber").Msg("vault decrypt failed")
			return common.ErrDecryption.WithContext(map[string]interface{}{
				"field": "mobileNumber",
			})
		}
		individual.MobileNumber = decrypted
	}

	if individual.AltContactNumber != "" {
		decrypted, err := s.vaultClient.Decrypt(ctx, individual.AltContactNumber, individual.TenantID)
		if err != nil {
			log.Error().Err(err).Ctx(ctx).Str("individualID", individual.ID).Str("field", "altContactNumber").Msg("vault decrypt failed")
			return common.ErrDecryption
		}
		individual.AltContactNumber = decrypted
	}

	for i := range individual.Identifiers {
		if individual.Identifiers[i].IdentifierType == common.IdentifierTypeAadhaar && individual.Identifiers[i].IdentifierID != "" {
			decrypted, err := s.vaultClient.Decrypt(ctx, individual.Identifiers[i].IdentifierID, individual.TenantID)
			if err != nil {
				log.Error().Err(err).Ctx(ctx).
					Str("individualID", individual.ID).
					Str("identifierType", individual.Identifiers[i].IdentifierType).
					Msg("vault decrypt failed on identifier")
				return common.ErrDecryption.WithContext(map[string]interface{}{
					"field": "identifierId",
					"type":  individual.Identifiers[i].IdentifierType,
				})
			}
			individual.Identifiers[i].IdentifierID = decrypted
		}
	}

	return nil
}

// EncryptIndividuals encrypts multiple individuals
func (s *encryptionService) EncryptIndividuals(ctx context.Context, individuals []models.Individual) error {
	for i := range individuals {
		if err := s.EncryptIndividual(ctx, &individuals[i]); err != nil {
			return err
		}
	}
	return nil
}

// DecryptIndividuals decrypts multiple individuals
func (s *encryptionService) DecryptIndividuals(ctx context.Context, individuals []models.Individual) error {
	for i := range individuals {
		if err := s.DecryptIndividual(ctx, &individuals[i]); err != nil {
			return err
		}
	}
	return nil
}

// HashMobileNumber computes the mobile-number blind index for search indexing.
func (s *encryptionService) HashMobileNumber(mobileNumber string) (string, error) {
	return common.HashMobileNumber(s.hmacSecret, mobileNumber), nil
}

File: conformance/code/individual/src/main/java/com/digit/individual/service/EncryptionService.java
package com.digit.individual.service;

import com.digit.individual.client.VaultClient;
import com.digit.individual.config.IndividualProperties;
import com.digit.individual.constants.ErrorCodes;
import com.digit.individual.model.Identifier;
import com.digit.individual.model.Individual;
import org.springframework.stereotype.Service;

import java.nio.charset.StandardCharsets;
import java.util.List;

/**
 * PII encryption/decryption. Mirrors Go internal/service/encryption_service.go.
 *
 * <p>Vault integration is config-gated ({@code individual.vault.enabled}). When disabled, encryption/
 * decryption are no-ops but the mobile number is still HMAC-SHA256 hashed into {@code hashedmobilenumber}
 * for indexed search — exactly the Go disabled path. When enabled, the Vault Transit engine encrypts
 * mobileNumber / altContactNumber / AADHAAR identifierIds (stored as {@code vault:v1:...} ciphertext)
 * while the keyed HMAC hash of the plaintext mobile number is stored alongside for exact-match search;
 * reads decrypt back to plaintext.
 */
@Service
public class EncryptionService {

    private final IndividualProperties.Vault config;
    private final VaultClient vaultClient;
    private final byte[] hmacSecret;

    public EncryptionService(IndividualProperties props, VaultClient vaultClient) {
        this.config = props.getVault();
        this.vaultClient = vaultClient;
        this.hmacSecret = props.getHmacSecret().getBytes(StandardCharsets.UTF_8);
        // Fail closed: when Vault is on the mobile column is encrypted at rest, so the blind index
        // MUST be keyed — otherwise the (reversible) hash would defeat the encryption. A blank pepper
        // here is a deploy misconfiguration, not something to silently accept. Mirrors Go config.Validate.
        if (config.isEnabled() && props.getHmacSecret().isEmpty()) {
            throw new IllegalStateException(
                    "HMAC_SECRET is required when Vault is enabled (mobile-number blind index must be keyed)");
        }
    }

    /** Encrypts PII fields in place. Mirrors Go encryptionService.EncryptIndividual. */
    public void encryptIndividual(Individual ind) {
        if (!config.isEnabled()) {
            // Still hash mobile number for search even without encryption.
            if (ind.getMobileNumber() != null && !ind.getMobileNumber().isEmpty()) {
                ind.setHashedMobileNumber(hashMobileNumber(ind.getMobileNumber()));
            }
            return;
        }

        // Encrypt mobile number (hash of the plaintext is retained for search).
        // Vault transit failures are genuine infra errors; let them propagate to the
        // tracer's generic 500 handler rather than masquerading as a business error.
        if (ind.getMobileNumber() != null && !ind.getMobileNumber().isEmpty()) {
            String plaintext = ind.getMobileNumber();
            String hash = hashMobileNumber(plaintext);
            String encrypted = vaultClient.encrypt(plaintext, ind.getTenantId());
            ind.setMobileNumber(encrypted);
            ind.setHashedMobileNumber(hash);
        }

        // Encrypt alt contact number.
        if (ind.getAltContactNumber() != null && !ind.getAltContactNumber().isEmpty()) {
            String encrypted = vaultClient.encrypt(ind.getAltContactNumber(), ind.getTenantId());
            ind.setAltContactNumber(encrypted);
        }

        // Encrypt identifiers (especially AADHAAR).
        List<Identifier> identifiers = ind.getIdentifiers();
        if (identifiers != null) {
            for (Identifier id : identifiers) {
                if (ErrorCodes.IDENTIFIER_TYPE_AADHAAR.equals(id.getIdentifierType())
                        && id.getIdentifierId() != null && !id.getIdentifierId().isEmpty()) {
                    String encrypted = vaultClient.encrypt(id.getIdentifierId(), ind.getTenantId());
                    id.setIdentifierId(encrypted);
                }
            }
        }
    }

    /** Decrypts PII fields in place. Mirrors Go encryptionService.DecryptIndividual. */
    public void decryptIndividual(Individual ind) {
        if (!config.isEnabled()) {
            return;
        }

        // Vault transit failures are genuine infra errors; let them propagate (tracer 500 handler).
        if (ind.getMobileNumber() != null && !ind.getMobileNumber().isEmpty()) {
            ind.setMobileNumber(vaultClient.decrypt(ind.getMobileNumber(), ind.getTenantId()));
        }

        if (ind.getAltContactNumber() != null && !ind.getAltContactNumber().isEmpty()) {
            ind.setAltContactNumber(vaultClient.decrypt(ind.getAltContactNumber(), ind.getTenantId()));
        }

        List<Identifier> identifiers = ind.getIdentifiers();
        if (identifiers != null) {
            for (Identifier id : identifiers) {
                if (ErrorCodes.IDENTIFIER_TYPE_AADHAAR.equals(id.getIdentifierType())
                        && id.getIdentifierId() != null && !id.getIdentifierId().isEmpty()) {
                    id.setIdentifierId(vaultClient.decrypt(id.getIdentifierId(), ind.getTenantId()));
                }
            }
        }
    }

    /**
     * Decrypts multiple individuals. Mirrors Go encryptionService.DecryptIndividuals, which returns on
     * the first row error; the search caller catches and logs it so the search still succeeds.
     */
    public void decryptIndividuals(List<Individual> individuals) {
        for (Individual ind : individuals) {
            decryptIndividual(ind);
        }
    }

    public String hashMobileNumber(String mobileNumber) {
        return HashUtil.hashMobileNumber(hmacSecret, mobileNumber);
    }
}
